### ISSUE TITLE
relay-runtime documentation confusion

### ISSUE THREAD

**Issue by magtastic** [facebook/relay#2611]
Hey! I am a lazy developer that just copies code examples from documentation with out much thought. I was "reading" the [environment setup guide](https://facebook.github.io/relay/docs/en/relay-environment.html) and saw the simple example. I use [this](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-extraneous-dependencies.md) eslint plugin, and I think many developers do (it is in the airbnb eslint preset).

As the lazy developer I am, I just threw one `yarn add relay-runtime` in the mix and everything was good again, no errors for me. Later on I started getting some strange errors, because of the my relay-runtime and relay version did not match, but I had no idea at the time. I spent some time debugging this and just figured this out...

Is there not some convention for some syntax like `require(relay/relay-runtime)` (or `require(relay/runtime)` for that matter), is there not?

**Comment by sibelius** [comment]
we should add relay-runtime as a required dependencies

right now it is installed from react-relay

**Comment by jstejada** [comment]
@sibelius, make it a required dependency of which other package? @magtastic, which package did you originally install?

**Comment by sibelius** [comment]
I think users should add relay-runtime explicit to their dependencies, without doing require('relay-runtime') won't work with yarn pnp and other tools

**Comment by magtastic** [comment]
@jstejada I just installed 
```
yarn add react react-dom react-relay
```
and
```
yarn add --dev babel-plugin-relay graphql
```

But when I was making the environment I saw this lint error and installed 
```
yarn add relay-runtime
```
The relay-runtime version did not match react-relay version so I got some strange errors.

**Comment by alloy** [comment]
So when you originally ran `yarn add relay-runtime` it installed v1.7.0 and when you ran `yarn add react react-dom react-relay` you got `react-relay` v2.0.0?

**Comment by Felix-N** [comment]
I had something similar yesterday.

I've updated a project to use 2.0 and ended up having not all dependencies updates.
I'm not sure how it happened though, I guess I did not save the package.json correctly once.

Annoying thing was figuring out what was wrong as everything run without errors - just having empty query parameters, as my ReactEnvironment could not resolve the query text.

In the end one dependency (probably the runtime) was back on 1.7. 
Some hint, that a dependency mismatch is occurring would be appreciated

**Comment by magtastic** [comment]
@alloy Well, actually I got react-relay v1.7.0 but `relay-runtime v2.0.0`. 

I just tried to reproduce this but got the following package.json
```
{
  "name": "blaa",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "react-relay": "^2.0.0",
    "relay-runtime": "^2.0.0"
  }
}
```

So I am guessing I might have installed `react-relay v.1.7.0` specifically, but honestly cannot remember what I did. (again, I am a lazy "developer" that copies and pastes most of my code and commands).

But the point still stands I think, if you do not intend people to add `relay-runtime` as a dependency, but use it through `react-relay`, the import path should be changed to something like `react-relay/runtime`, and if `relay-runtime` should be installed separately it should have `react-relay` in peerDependencies. Right?

**Comment by alloy** [comment]
> Well, actually I got `react-relay v1.7.0` but `relay-runtime v2.0.0`.

I see. I guess the botched release of v2 referred to by @josephsavona [here](https://github.com/facebook/relay/issues/2615#issuecomment-457845453) must have successfully released `relay-runtime v2.0.0`, but _not_ `react-relay v2.0.0` and this must have been rectified in the meantime.

> if you do not intend people to add `relay-runtime` as a dependency, but use it through `react-relay`, the import path should be changed to something like `react-relay/runtime`

I don't think it's the case that you shouldn't use `relay-runtime` directly (in any case we use it directly), what makes you think that? Or why a shim would be better?

> if `relay-runtime` should be installed separately it should have `react-relay` in peerDependencies.

`relay-runtime` is supposed to be view framework agnostic, so it shouldn't in any way depend on `react-relay`.

----

All in all, I think your experience may mostly have been a very unfortunate situation of the release of v2.0.0 not going smoothly and am not sure what could be done to really improve this edge-case for the future (although I'm jetlagged and may just be missing something obvious!).

**Comment by jstejada** [comment]
Hmm weird, afaict all packages should’ve been released to 2.0 at the same time. I’ll try to get a repro.

Let me double check the dependencies and the state of the packages right now

**Comment by jstejada** [comment]
I ran the same steps @magtastic mentioned in this is the result:

```
yarn add react react-dom react-relay
```

![image](https://user-images.githubusercontent.com/1271509/51913905-7497a680-238c-11e9-9a20-cc0c96df2331.png)

```
yarn add --dev babel-plugin-relay graphql
```

![image](https://user-images.githubusercontent.com/1271509/51913931-87aa7680-238c-11e9-9e41-438866acbf73.png)


so I'm not sure how you managed to get mismatching versions

**Comment by alloy** [comment]
> But as @Felix-N pointed out, it would be nice to add some kind of warning somewhere if you are using react-relay and relay-runtime together with incompatible versions.

Yeah it would be, but not sure what that would be. Node is made for you to be able to have multiple versions of the same package, so maybe this should be handled at the dependency manager layer? E.g. perhaps adding a peer-dependency on relay-runtime _besides_ the hard runtime-dependency would trigger the dependency manager to warn about your package pulling in a incompatible version? 🤔

**Comment by Felix-N** [comment]
I can only comment form a user persepctive (i'm not deep enough in the details of the package).

In my case something apparently wen't wrong and the wrong version was installed. What followed was that my app did not throw any error at all, but could not fetch data anymore. Reason was, that the query parameter was always empty. At that moment it was a bit tricky to figure out what was happening. Basically my ReactEnvironment didn't get the operation.text anymore, but suddenly operation.params.text
Changing that resulted in normal queries to work, but I'd break other parts with it in relay.
In the end I figured out, that one lib did not update to 2.0 but stayed on 1.7 - again I'm not sure what the reason for that was. Most likely an error from me.

Never the less, finding the problem without any hint whats wrong, was a bit annoying :)
